File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    bool isValid(vector<vector<char>>& grid, vector<vector<bool>>& visited, int x, int y) {
        if(x < 0 || x >= grid[0].size() || y < 0 || y >= grid.size() || visited[y][x] || grid[y][x] == '0') return false;
        return true;
    }

    void dfs(vector<vector<char>>& grid, vector<vector<bool>>& visited, int x, int y) {
        visited[y][x] = true;
        if(isValid(grid, visited, x + 1, y)) dfs(grid, visited, x + 1, y);
        if(isValid(grid, visited, x - 1, y)) dfs(grid, visited, x - 1, y);
        if(isValid(grid, visited, x, y + 1)) dfs(grid, visited, x, y + 1);
        if(isValid(grid, visited, x, y - 1)) dfs(grid, visited, x, y - 1);
    }

    int numIslands(vector<vector<char>>& grid) {
        int x = grid[0].size();
        int y = grid.size();
        vector<vector<bool>> visited(y, vector<bool>(x, false));
        int tot = 0;
        for(int i = 0; i < y; i++) {
            for(int j = 0; j < x; j++) {
                if(!visited[i][j] && grid[i][j] == '1') {
                    tot++;
                    dfs(grid, visited, j, i);
                }
            }
        }

        return tot;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.ts
function numIslands(grid: string[][]): number {
    let m = grid.length, n = grid[0].length;

    const visited: boolean[][] = Array.from({ length: m }, () => new Array(n).fill(false));

    const xArr = [1, 0, -1, 0];
    const yArr = [0, 1, 0, -1];

    function bfs(x, y) {
        const q: [number, number][] = [[x, y]];
        visited[x][y] = true;

        while (q.length > 0) {
            const [row, col] = q.shift();

            for (let i = 0; i < 4; i++) {
                const newX = row + xArr[i], newY = col + yArr[i];

                if (isValid(newX, newY)) {
                    visited[newX][newY] = true;
                    q.push([newX, newY]);
                }
            }
        }
    }

    function isValid(newX: number, newY: number): boolean {
        if (newX < 0 || newX >= m || newY < 0 || newY >= n || visited[newX][newY] || grid[newX][newY] === "0") return false;
        return true;
    }

    let ans = 0;
    for(let i = 0; i < m; i++) {
        for(let j = 0; j < n; j++) {
            if(grid[i][j] === "1" && !visited[i][j]) {
                bfs(i, j);
                ans++;
            }
        }
    }

    return ans;
};